spec/registry.md

```diff
@@ -132,14 +136,15 @@ when * {{{$count :number} new messages}}
 
 Furthermore,
 `:number`'s `<matchSignature>` contains two `<match>` elements
-which allow to validate the variant keys.
-If at least one `<match>` validation rules passes,
-a variant key is considered valid.
-
-- `<match validationRule="anyNumber"/>` can be used to valide the `when 1` variant
+which allow the validation of variant keys.
+The first element with `locales` matching the current locale
```

Comment:
The bullet item below says "can be used in locales like" and this says "matching" in a pretty non-specific way. That's probably not helpful to registry authors. We should specify the matching scheme (lookup is mechanical, best match is what some will want)

Reply:
Probably lookup? Is there an explicit reference here we can use for that?

Reply:
BCP47 lookup is mechanical, but often produces very bad results. (Eg mixture of sr_Latn with sr_Cyrl in the same message.

Also, the matching is really bad, due to cramming values that are both for ordinal and cardinal. The valid values are going to vary depending on the options.
```
            <match locales="en" values="one two few other" validationRule="anyNumber"/>
            <match values="zero one two few many other" validationRule="anyNumber"/>
```

Moreover, it would be repeating data in CLDR, a really bad idea. It should be possible in the registry to reference external sources.

Reply:
```suggestion
The first element with `locales` matching the current locale
according to the [BCP 47 Lookup](https://datatracker.ietf.org/doc/html/rfc4647#section-3.4) scheme
```

Reply:
For context, I'm not at all concerned about JS `Intl` here, but about much more limited environments. For a real-world example, Firefox's crash reporter is a separate executable that does not have access to the ICU that Firefox comes with, or CLDR data. Its localization will need to be wholly self-contained. It would be Very Useful to be able to still use MF2 for its localization.

As it's currently expressed, MF2 fits that bill. But if we add here a requirement to use a better matching algorithm than Lookup that e.g. knows that `sr` and `sr-Cyrl` are synonymous, we're effectively introducing a dependency on the CLDR that I'd rather we didn't.

And that's especially as I very strongly suspect that all real-world `<match>` sets will be rather reasonable ones, and so will let a mechanical first-match scheme perform just as well as any other scheme would.

Reply:
I hear you. FWIW, I'm a fan of BCP47 Lookup's mechanical approach for situations like this (resource files and the like) where you want predictable results across many implementations. I called out JS `Intl`, though, because it offers both options. The "best fit" option there is described by MDN as "provid[ing] a locale that's at least, but possibly more, suited for the request than the result of the Lookup algorithm". Allowing for both might be a good idea, if some implementations end up using ICU-based functionality under the hood.

I'm engaged in publishing some more material about this following today's W3C call. [This text](https://w3c.github.io/localizable-manifests/#locale-fallback) might also eventually appear in [LTLI](https://www.w3.org/TR/ltli) as well as other places.

For this text, I would suggest the following changes. Note that this also omits "first" from "first element...matching" since matching does not stop with the first potential match.

```suggestion
The element whose `locales` best matches the current locale is used. 
The "best match" is determined according to [BCP47 Lookup](https://datatracker.ietf.org/doc/html/rfc4647#section-3.4) 
or an appropriately-compatible implementation-specific "best fit" scheme
such as that [found in Javascript](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Intl#locale_identification_and_negotiation)
or used by CLDR.
An element with no `locales` attribute is considered the default.
```

(Also, note that the next line in your current text misspells `element`)

Reply:
I disagree. It would be better to have no algorithm, and have people explicitly add the locales they want to match, then it would be to get the wrong matches. And flipping scripts can have really bad ransom-note effects.

Example:
```
 // if you really don't care which script the output is in
<match ... locales="sr sr-Cyrl sr-Latn">

// otherwise either 
<match ... locales="sr sr-Cyrl">
// or
<match ... locales="sr sr-Latn">
```

Reply:
Allowing for optionality here would mean that two implementations could end up selecting different rules when given the same registry and locale. Could we not rely on registry definitions being reasonable, and accept the first match for which Lookup succeeds?

A validator/linter for the registry XML itself could wholly separately point out to its author if it contains any mis-ordering that would be problematic.

Reply:
@macchiati noted:
>  It would be better to have no algorithm, and have people explicitly add the locales they want to match, then it would be to get the wrong matches.

This would effectively require listing *every* locale in the registry (even if we allow the current locale to have its extensions trimmed off), since there would be no prefix matching. That doesn't seem reasonable?

`plural` isn't a good example from the point of view that we want to point at the CLDR data rather than mirroring it into the registry. However, it does provide a reasonable example of what an arbitrary function would need. The CLDR data for plural rules does not list every locale. It inherits down (usually from the primary language subtag) in most cases (including, I would point out, the `sr` one). Obviously cross-script cases and others require greater care (although it doesn't affect plural rules in general).

An alternative here would be to not provide for the registry to specify locale-based variations. Tools that want to do things such as matrix explosion will then need out-of-band information. Using `plural` as an example, the variation list would be:

_anyNumber_, `zero`, `one`, `two`, `few`, `many`, `other`, (implied `*`)

It would be valid to write any locale _message_ using those keys (and, in fact, I have done this for I18N demos, particularly in the root locale). A developer writing a message and assuming the `en` locale will use keys `0`/`one`/`*`; they would want translators to expand the key set according to the target locale (so Polish ends up with `0`/`one`/`few`/`many`/`*`). The registry, though, would not help them.

For a developer writing a custom function, they won't have CLDR data to help them and so would potentially use @eemeli's feature to create this information. That way machines (CAT tools, not runtimes) could read it to generate the target language keyset. If the locale I'm creating a resource for is `sr-Latn-BS` and the registry has root and `sr`, I probably want the `sr` set and not the root.

---
@eemeli 
> accept the first match for which Lookup succeeds?

It should not be the *first* match. It should always be the *best* match. Note well that "best" is overloaded here.

Consider this example:

```
"localized-resource": {
	"en":     {"value": "This is English"},
	"en-GB":  {"value": "This is UK English"},
	"en-US":  {"value": "This is US English"}
}
```

For the locale `en-GB-oxendict`, `en` is the *first* match, but it is not the *best* match. `en-GB` is the *best* match. (This is pure Lookup, please note). A "best fit" matcher could decide that a request for `br` (Breton) matches `fr` before falling back to root (to use one of Mark's favorite examples) and that would not be BCP47 Lookup.

Reply:
> Consider this example:
> 
> ```
> "localized-resource": {
> 	"en":     {"value": "This is English"},
> 	"en-GB":  {"value": "This is UK English"},
> 	"en-US":  {"value": "This is US English"}
> }
> ```
> 
> For the locale `en-GB-oxendict`, `en` is the _first_ match, but it is not the _best_ match. `en-GB` is the _best_ match. (This is pure Lookup, please note). A "best fit" matcher could decide that a request for `br` (Breton) matches `fr` before falling back to root (to use one of Mark's favorite examples) and that would not be BCP47 Lookup.

Ah! I was not previously clear in what I meant by e.g. "reasonable" when describing a registry. Apologies.

In the XML, we don't have a mapping as above, but an ordered sequence of e.g. `<match>` statements. If the Lookup is done by selecting first-match, then the "reasonable" order for the ones you list would be:

```xml
<match locales="en-GB"/>
<match locales="en-US"/>
<match locales="en"/>
```

With that order, a first-match selection would produce the same results as best-match, because the prioritisation ends up being encoded positionally in the XML.

Reply:
At the cost of requiring the tags to be ordered. Moreover, there are sets of tags that might be difficult to order (e.g. `zh-CN` vs. `zh-Hans` vs. `zh-Hans-CN`). There are other ways in XML to perform selection which might be more appropriate here.

---

Part of me is still puzzled by the need to represent locales in the registry in this way. I think that a `match` element represents _potential_ output values of a selector (in a specific locale set when it has `locales`). It doesn't represent the specific outputs (which depend on evaluating the inputs). So it isn't consumed by the runtime (which would just call the function anyway). Instead it describes the potential outputs for cases such as I describe above (CAT tools exploding or contracting the variant matrix). 

Is that what you're thinking this is for? If that's the case, is that "nice to have" for release? Or do you think there will be a lot of cases where there are selectors whose potential output varies outside a closed set for different locales, e.g.

(default): `a`, `b`, `c`
en: `b`, `c`, `d`
fr: `e`, `f`, `g`

Such that matrices look different for a given match statement in a message. 

```
.match {$var :function}
[a] {{Default locale can get here}}
[b] {{default or en}}
[c] {{default or en}}
[d] {{en only}}
[e] {{fr only}}
[f] {{fr only}}
[g] {{fr only}}
[*] {{everyone can get here}}
```

Reply:
> At the cost of requiring the tags to be ordered. Moreover, there are sets of tags that might be difficult to order (e.g. `zh-CN` vs. `zh-Hans` vs. `zh-Hans-CN`). There are other ways in XML to perform selection which might be more appropriate here.

The tricky bit here is of course what to do if we have
```xml
<match locales="zh-CN"/>
<match locales="zh-Hans"/>
```
and our current locale is `zh-Hans-CN`. And I'd argue that with first-match it's actually easier to resolve this than with best-match, as the ordering can define for this selector in particular whether to prioritise the script or the region.

In practice, I would be surprised if any real-world selectors turn out to depend not only on the primary subtag, but also the region _and_ script. For example, with plurals the _only_ one with any variance beyond the primary is `pt-PT` vs. `pt`.

> Part of me is still puzzled by the need to represent locales in the registry in this way. [...] Instead it describes the potential outputs for cases such as I describe above (CAT tools exploding or contracting the variant matrix).
> 
> Is that what you're thinking this is for? If that's the case, is that "nice to have" for release? [...]

This has no impact on formatting, correct, but the improvement we're offering for tooling is significant enough that if we get this right, it may by itself end up as a reason for promoting migration to MF2. At the moment, no data definition like this exists for any format. So if e.g. a CAT tool wants to be able to map plural categories correctly when translating, it needs to figure out a custom way of getting the data for that, and applying it.

Instead, here we're defining a regular shape to such data, and not just for plurals but any other selector as well. And not just to proprietary CAT tools, but _all_ such tools so that _any_ validator based on MF2 registry data can validate the result coming from _any_ CAT tool.

I think of what's proposed here as a minimum we'll need for the release. There are further aspects that I think we'll want to follow up with possibly post-release that are "nice to have", such as being able to encode data for the intersection of "integer value" and "plural selector", or how/when leaving out the `many` plural case for Romance languages is appropriate. On that note: we ought to ensure that we can version the registry definition separately from the spec, so that it can experience breaking changes.

Reply:
> Instead, here we're defining a regular shape to such data, and not just for plurals but any other selector as well. And not just to proprietary CAT tools, but all such tools so that any validator based on MF2 registry data can validate the result coming from any CAT tool.

Okay, so we're in agreement about the objective here, I think.

Yes, I agree that MF2 processors such as CAT tools or resource editors or other parts of the ecosystem should be able to have a machine-readable way of figuring out how to manage variant sets that vary by locale. 

For cases in MF2's default registry, I believe that the source data for such information will come from existing CLDR data (such as the plural rules!) and that we should provide an interface to that data.

If the existing CLDR data cannot be referenced in-situ and a transform is required, we should include that transform in our release and ensure that each new CLDR release automagically updates the MF2 registry. **This means the MF2 default registry will be versioned with CLDR.** (This is not necessarily a bad thing)

What you're defining here (to go back on topic) is the registry structure for selectors (including custom ones written by users of a given implementation). The test of this would be our ability to transform CLDR data for built-in selectors to deliver that format.

Assuming that this is correct, the locale matching we specify probably needs to exactly match that used by CLDR. As you note, plural rules are not exactly a deep and exacting test of this. Ditto ordinals. Maybe that's a good thing too. In any case, we can can specify how `<match locales=...>` is matched by referencing LDML [roughly here](https://unicode.org/reports/tr35/#Bundle_vs_Item_Lookup). This is a higher bar than BCP47 Lookup, but in practice it is extremely similar. The main problem is `addLikelySubtags`, which can be implemented with a small amount of code and about 14k of (uncompressed) textual data in my implementation (I didn't spend any effort on optimizing for size). Most locales won't be affected. In fact, **_we_** won't need it for the default registry values we produce. There can be a test or tests for "registry consumers", by which we do not mean MF2 formatter implementations. Those will just call the selector.

Reply:
Okay, you're right.

Thinking this through again, I'd not accounted for the registry being completely irrelevant during formatting, where the resource constraints are tight. Let us indeed use CLDR matching exactly. That'll also obviate the need for any plural data transforms to re-order their results, which could be quite hard to express in XSLT.

Could you suggest language here to that effect?

Reply:
```suggestion
The element whose `locales` best matches the current locale 
using resource item [lookup](https://unicode.org/reports/tr35/#Lookup) from LDML is used.
An element with no `locales` attribute is the default 
(and is considered equivalent to the `root` locale).
```